src/code/z_lifemeter.c

```diff
@@ -1,17 +1,434 @@
+                }
+            } else if (i > fullHeartCount) {
+                if (curColorSet != 6) {
+                    curColorSet = 6;
+                    gDPPipeSync(OVERLAY_DISP++);
+                    gDPSetPrimColor(OVERLAY_DISP++, 0, 0, sHeartsDDPrim[0][0], sHeartsDDPrim[0][1], sHeartsDDPrim[0][2],
+                                    interfaceCtx->healthAlpha);
+                    gDPSetEnvColor(OVERLAY_DISP++, sHeartsDDEnv[0][0], sHeartsDDEnv[0][1], sHeartsDDEnv[0][2], 255);
+                }
+            } else if (curColorSet != 7) {
+                curColorSet = 7;
+                gDPPipeSync(OVERLAY_DISP++);
+                gDPSetPrimColor(OVERLAY_DISP++, 0, 0, sHeartsDDPrim[1][0], sHeartsDDPrim[1][1], sHeartsDDPrim[1][2],
+                                interfaceCtx->healthAlpha);
+                gDPSetEnvColor(OVERLAY_DISP++, sHeartsDDEnv[1][0], sHeartsDDEnv[1][1], sHeartsDDEnv[1][2], 255);
+            }
+            if (i < fullHeartCount) {
+                heartTex = &D_02000900;
+            } else if (fullHeartCount == i) {
+                heartTex = HeartDDTextures[fractionHeartCount];
+            } else {
+                heartTex = &D_02000500;
+            }
+        }
+
+        if (temp != heartTex) {
+            temp = heartTex;
+            gDPLoadTextureBlock(OVERLAY_DISP++, heartTex, G_IM_FMT_IA, G_IM_SIZ_8b, 16, 16, 0,
+                                G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMASK, G_TX_NOMASK,
+                                G_TX_NOLOD, G_TX_NOLOD);
+        }
+
+        if (fullHeartCount != i) {
+            if ((ddCount < 0) || (i > ddCount)) {
+                if (curCombineModeSet != 1) {
+                    curCombineModeSet = 1;
+                    func_8012C654(gfxCtx);
+                    gDPSetCombineLERP(OVERLAY_DISP++, PRIMITIVE, ENVIRONMENT, TEXEL0, ENVIRONMENT, TEXEL0, 0, PRIMITIVE,
+                                      0, PRIMITIVE, ENVIRONMENT, TEXEL0, ENVIRONMENT, TEXEL0, 0, PRIMITIVE, 0);
+                }
+            } else if (curCombineModeSet != 3) {
+                curCombineModeSet = 3;
+                func_8012C654(gfxCtx);
+                gDPSetCombineLERP(OVERLAY_DISP++, ENVIRONMENT, PRIMITIVE, TEXEL0, PRIMITIVE, TEXEL0, 0, PRIMITIVE,
+                                    0, ENVIRONMENT, PRIMITIVE, TEXEL0, PRIMITIVE, TEXEL0, 0, PRIMITIVE, 0);
+            }
+            posY = 26.0f + offsetY;
+            posX = 30.0f + offsetX;
+            temp_f4 = 1.0f;
+            temp_f4 /= 0.68f;
+            temp_f4 *= 1 << 10;
+            halfTexSize = 8.0f;
+            halfTexSize *= 0.68f;
+            gSPTextureRectangle(OVERLAY_DISP++, (s32)((posX - halfTexSize) * 4), (s32)((posY - halfTexSize) * 4),
+                                (s32)((posX + halfTexSize) * 4), (s32)((posY + halfTexSize) * 4), G_TX_RENDERTILE,
+                                0, 0, (s32)temp_f4, (s32)temp_f4);
```

Comment:
```suggestion
            gSPTextureRectangle(OVERLAY_DISP++, (posX - halfTexSize) * 4, (posY - halfTexSize) * 4,
                                (posX + halfTexSize) * 4, (posY + halfTexSize) * 4, G_TX_RENDERTILE,
                                0, 0, temp_f4, temp_f4);
```
Are these `(s32)` casts needed?

Reply:
yes those cast are needed to match